qcodes/instrument_drivers/yokogawa/GS200.py

```diff
@@ -191,38 +191,43 @@ def __init__(self, name: str, address: str, terminator: str = "\n",
                            set_cmd=self._set_source_mode,
                            vals=Enum('VOLT', 'CURR'))
 
-        # When getting the mode internally in the driver, look up the mode as
-        # recorded by the _cached_mode property, instead of calling source
-        # _mode(). This will prevent frequent VISA calls to the instrument.
-        # Calling _set_source_mode will change the chased value.
-        self._cached_mode = "VOLT"
+        # We need to get the source_mode value here as we cannot rely on the
+        # default value that may have been changed before we connect to the
+        # instrument (in a previous session or via the frontpanel).
+        self.source_mode()
+
+        self.add_parameter('range',
+                           label='Source Range',
+                           get_cmd=':SOUR:RANG?',
+                           set_cmd=partial(self._set_range, self.source_mode()),
+                           vals=Enum(1e-3, 10e-3, 100e-3, 200e-3, 1e0,
+                           10e0, 30e0),
+                           )
```

Comment:
I don't think that it is a good idea to add this range parameter. Now there are two different parameters that reprecent the same value (In voltage mode voltage_range and range) furthermore this does not correctly do the logic that voltage and current range does

Reply:
The only reason for adding this range parameter is to use it in _set_output method. This method checks if we are in correct range to set the output value and raise error if the value is out of range. In the original code, a local _cache_range_value served this checking, which we wanted to get rid of that. So, what do you suggest for the replacement of _cache_range_value?

Reply:
You are confusing two different things.

`self.range` is an alias for either `self.voltage_range` or self.current_range` depending on the mode. You might think that is it safe to replace a alias with a new definition of the parameter with the same command but it is not since the snapshot will be out of date e.g. if you set the voltage_range the snapshot of range will not update and the other way around. 

Futher to that there was an internal cache of the range value stored in _cache_range_value. That is unneeded and error prone since both `voltage_range` and `current_range` have their own internal cache and the alias to them called `range` therefor also does.

Rather than just creating an alias as the orginal code does it would be better to use a DelegateParameter https://qcodes.github.io/Qcodes/api/parameters/parameter.html?qcodes.instrument.parameter.DelegateParameter#qcodes.instrument.parameter.DelegateParameter for `range` and change its source as needed

Reply:
> self.range is an alias for either self.voltage_range or self.current_range
True. The problem of this approach is that it always assumes we initialize the instrument in the voltage mode.

> Futher to that there was an internal cache of the range value stored in _cache_range_value...
This is fair.

> Rather than just creating an alias as the orginal code does it would be better to use a DelegateParameter...
I will try this method